opt: exec.Factory cleanup

RaduBerinde opened the pull request:
#### opt: better type safety for buffer nodes in exec.Factory

This change adds an `exec.BufferNode` interface, which makes it more clear where
only such nodes can be used.

Release note: None

#### exec: separate node and table column ordinals in exec.Factory

We use ColumnOrdinal / ColumnOrdinalSet in the factory to refer to either table
column ordinals or node column ordinals. This can be confusing and has led to
bugs in the past. This commit separates the two types.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47736)
<!-- Reviewable:end -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47736)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/exec/factory.go, line 599 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47736#-M5PxHiC-9nLPlfpRO7a:-M5PxHiC-9nLPlfpRO7b:bx55bac) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3340b3a52959ff4a7e1812faa84df27bb88138/pkg/sql/opt/exec/factory.go#L599)):*
> ```Go
> )
> 
> // TableColumnOrdinal is the 0-based ordinal index of a cat.Table column.
> ```

Can you expand this and/or the `NodeColumnOrdinal` comment to better explain them? In particular, what's an example case where the two could get confused and cause a bug?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47736)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @rytaft)

___
*[pkg/sql/opt/exec/factory.go, line 599 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47736#-M5PxHiC-9nLPlfpRO7a:-M5RsCJq49Bot2wCHO40:bigfyws) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3340b3a52959ff4a7e1812faa84df27bb88138/pkg/sql/opt/exec/factory.go#L599)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Can you expand this and/or the `NodeColumnOrdinal` comment to better explain them? In particular, what's an example case where the two could get confused and cause a bug?
</blockquote></details>

Mutations and lookup/index joins are examples where we have both a Node input and a table and things can get confusing when all arguments are just `ColumnOrdinal`. I remember we had a bug somewhere but can't remember what it was exactly (maybe I had it during development).

Expanded the comments a bit.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 7 files at r1, 6 of 6 files at r3, 7 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47736#-:-M5Sd0Fm80lnOk29kN24:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball)



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47736)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @RaduBerinde)

___
*[pkg/sql/buffer.go, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47736#-M5Sny9h2iSruf7p9Raf:-M5Sny9h2iSruf7p9Rag:b93l9a5) ([raw file](https://github.com/cockroachdb/cockroach/blob/86cfd2497f6c66d30166db51fba7c37799ce9a55/pkg/sql/buffer.go#L38)):*
> ```Go
> 
> // BufferNodeMarker is part of the exec.BufferNode interface.
> func (n *bufferNode) BufferNodeMarker() {}
> ```

I don't understand the need for this no-op function.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/buffer.go, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47736#-M5Sny9h2iSruf7p9Raf:-M5SoTOBDDOKJtXk7c_t:bhhnel4) ([raw file](https://github.com/cockroachdb/cockroach/blob/86cfd2497f6c66d30166db51fba7c37799ce9a55/pkg/sql/buffer.go#L38)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I don't understand the need for this no-op function.
</blockquote></details>

Without it, any type would implement `exec.BufferNode`, it would be no different than `exec.Buffer`.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47736)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @RaduBerinde)

___
*[pkg/sql/buffer.go, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47736#-M5Sny9h2iSruf7p9Raf:-M5SopaX0dAsmsO1jLv4:b-jhyyta) ([raw file](https://github.com/cockroachdb/cockroach/blob/86cfd2497f6c66d30166db51fba7c37799ce9a55/pkg/sql/buffer.go#L38)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Without it, any type would implement `exec.BufferNode`, it would be no different than `exec.Buffer`.
</blockquote></details>

Got it. Thanks!


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: better type safety for buffer nodes in exec.Factory

This change adds an `exec.BufferNode` interface, which makes it more clear where
only such nodes can be used.

Release note: None

RaduBerinde committed:
exec: separate node and table column ordinals in exec.Factory

We use ColumnOrdinal / ColumnOrdinalSet in the factory to refer to either table
column ordinals or node column ordinals. This can be confusing and has led to
bugs in the past. This commit separates the two types.

Release note: None

RaduBerinde commented:
bors r+
